pkg/smokescreen/conntrack/conn_tracker.go

```diff
@@ -15,19 +17,92 @@ type Tracker struct {
 	Wg           *sync.WaitGroup
 	statsc       statsd.ClientInterface
 
+	SuccessRateTracker *ConnSuccessRateTracker
+
 	// A connection is idle if it has been inactive (no bytes in/out) for this
 	// many seconds.
 	IdleTimeout time.Duration
 }
 
-func NewTracker(idle time.Duration, statsc statsd.ClientInterface, logger *logrus.Logger, sd atomic.Value) *Tracker {
+// ConnSuccessRateTracker tracks statistics about the overall success rate
+// of connection attempts over some time interval (which is set in StartNewConnSuccessRateTracker()).
+//
+// It tracks only the *most recently seen* connection to an individual destination host within the configured
```

Comment:
Also wondering what normalization you think might be appropriate. If the population of outgoing requests looks like
```
customer-4355a46b19.bighostingsite.com
customer-53c234e5e8.bighostingsite.com
customer-1121cfccd5.bighostingsite.com
customer-7de1555df0.bighostingsite.com
customer-f0b5c2c221.bighostingsite.com
customer-06e9d52c17.bighostingsite.com
customer-10159baf26.bighostingsite.com
customer-aa67a169b0.bighostingsite.com
blorp.lol
customer-917df3320d.bighostingsite.com
```
and `bighostingsite.com` is completely down, it will look like a 90% failure rate.

Reply:
eTLD+1? There's a nice [go library](https://pkg.go.dev/golang.org/x/net/publicsuffix) for this, I wrote a function to normalize domains below, but let me know if it seems like there are cases I haven't thought of